flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala

```diff
@@ -0,0 +1,188 @@
+import org.apache.flink.api.table.validate.ValidationException
+
+case class Project(projectList: Seq[NamedExpression], child: LogicalNode) extends UnaryNode {
+  override def output: Seq[Attribute] = projectList.map(_.toAttribute)
+
+  override def toRelNode(relBuilder: RelBuilder): RelBuilder = {
+    def allAlias: Boolean = {
+      projectList.forall { proj =>
+        proj match {
+          case Alias(r: ResolvedFieldReference, name) => true
+          case _ => false
+        }
+      }
+    }
+    child.toRelNode(relBuilder)
+    if (allAlias) {
+      relBuilder.push(
+        LogicalProject.create(relBuilder.peek(),
+          projectList.map(_.toRexNode(relBuilder)).asJava,
+          projectList.map(_.name).asJava))
+    } else {
+      relBuilder.project(projectList.map(_.toRexNode(relBuilder)): _*)
+    }
+  }
+}
+
+case class AliasNode(aliasList: Seq[Expression], child: LogicalNode) extends UnaryNode {
+  override def output: Seq[Attribute] =
+    throw new UnresolvedException("Invalid call to output on AliasNode")
+
+  override def toRelNode(relBuilder: RelBuilder): RelBuilder =
+    throw new UnresolvedException("Invalid call to toRelNode on AliasNode")
+
+  override lazy val resolved: Boolean = false
+}
+
+case class Distinct(child: LogicalNode) extends UnaryNode {
+  override def output: Seq[Attribute] = child.output
+
+  override def toRelNode(relBuilder: RelBuilder): RelBuilder = {
+    child.toRelNode(relBuilder)
+    relBuilder.distinct()
+  }
+}
+
+case class Filter(condition: Expression, child: LogicalNode) extends UnaryNode {
+  override def output: Seq[Attribute] = child.output
+
+  override def toRelNode(relBuilder: RelBuilder): RelBuilder = {
+    child.toRelNode(relBuilder)
+    relBuilder.filter(condition.toRexNode(relBuilder))
+  }
+}
+
+case class Aggregate(
+    groupingExpressions: Seq[Expression],
+    aggregateExpressions: Seq[NamedExpression],
+    child: LogicalNode) extends UnaryNode {
+
+  override def output: Seq[Attribute] = {
+    (groupingExpressions ++ aggregateExpressions) map { agg =>
+      agg match {
+        case ne: NamedExpression => ne.toAttribute
+        case e => Alias(e, e.toString).toAttribute
+      }
+    }
+  }
+
+  override def toRelNode(relBuilder: RelBuilder): RelBuilder = {
+    child.toRelNode(relBuilder)
+    relBuilder.aggregate(
+      relBuilder.groupKey(groupingExpressions.map(_.toRexNode(relBuilder)).asJava),
+      aggregateExpressions.filter(_.isInstanceOf[Alias]).map { e =>
+        e match {
+          case Alias(agg: Aggregation, name) => agg.toAggCall(name)(relBuilder)
+          case _ => null // this should never happen
+        }
+      }.asJava)
+  }
+}
+
+case class Union(left: LogicalNode, right: LogicalNode) extends BinaryNode {
+  override def output: Seq[Attribute] = left.output
+
+  override def toRelNode(relBuilder: RelBuilder): RelBuilder = {
+    left.toRelNode(relBuilder)
+    right.toRelNode(relBuilder)
+    relBuilder.union(true)
+  }
+}
```

Comment:
Need to override the `resolved` field to check equality of `left` and `right` output schema. Will do in next commit.